core/services/directrequestocr/orm.go

```diff
@@ -19,121 +20,98 @@ type ORM interface {
-	newEntry := Request{
-		ID:            o.counter,
-		RequestID:     requestID,
-		ReceivedAt:    receivedAt,
-		RequestTxHash: requestTxHash,
-		State:         IN_PROGRESS,
-	}
-	o.db[requestID] = newEntry
-	return nil
+func (o orm) SetResult(requestID RequestID, runID int64, computationResult []byte, readyAt time.Time, qopts ...pg.QOpt) error {
+	stmt := `
+		UPDATE ocr2dr_requests
+		SET run_id=$3, result=$4, result_ready_at=$5, state=$6
+		WHERE request_id=$1 AND contract_address=$2;
+	`
+	return o.q.WithOpts(qopts...).ExecQ(stmt, requestID[:], o.contractAddress.Bytes(), runID, computationResult, readyAt, RESULT_READY)
 }
 
-func (o *inmemoryorm) SetResult(requestID RequestID, runID int64, computationResult []byte, readyAt time.Time, qopts ...pg.QOpt) error {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	if val, ok := o.db[requestID]; ok {
-		val.RunID = runID
-		val.ErrorType = NONE
-		val.Result = computationResult
-		val.Error = []byte{}
-		val.State = RESULT_READY
-		val.ResultReadyAt = readyAt
-		o.db[requestID] = val
-		return nil
-	}
-	return fmt.Errorf("can't find entry with requestID: %v", requestID)
+func (o orm) SetError(requestID RequestID, runID int64, errorType ErrType, computationError []byte, readyAt time.Time, qopts ...pg.QOpt) error {
+	stmt := `
+		UPDATE ocr2dr_requests
+		SET run_id=$3, error=$4, error_type=$5, result_ready_at=$6, state=$7
+		WHERE request_id=$1 AND contract_address=$2;
+	`
+	return o.q.WithOpts(qopts...).ExecQ(stmt, requestID[:], o.contractAddress, runID, computationError, errorType, readyAt, RESULT_READY)
 }
 
-func (o *inmemoryorm) SetError(requestID RequestID, runID int64, errorType ErrType, computationError []byte, readyAt time.Time, qopts ...pg.QOpt) error {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	if val, ok := o.db[requestID]; ok {
-		val.RunID = runID
-		val.ErrorType = errorType
-		val.Error = computationError
-		val.State = RESULT_READY
-		val.Result = []byte{}
-		val.ResultReadyAt = readyAt
-		o.db[requestID] = val
-		return nil
+func (o orm) SetState(requestID RequestID, state RequestState, qopts ...pg.QOpt) (RequestState, error) {
+	var oldState RequestState
+
+	stmt := `SELECT state FROM ocr2dr_requests WHERE request_id=$1 AND contract_address=$2;`
+	if err := o.q.WithOpts(qopts...).Get(&oldState, stmt, requestID[:], o.contractAddress); err != nil {
+		return 0, err
 	}
-	return fmt.Errorf("can't find entry with requestID: %v", requestID)
-}
 
-func (o *inmemoryorm) SetState(requestID RequestID, state RequestState, qopts ...pg.QOpt) (RequestState, error) {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	prevState := IN_PROGRESS
-	if val, ok := o.db[requestID]; ok {
-		prevState = val.State
-		val.State = state
-		o.db[requestID] = val
-		return prevState, nil
+	if oldState == state {
+		return state, nil
 	}
-	return prevState, fmt.Errorf("can't find entry with requestID: %v", requestID)
+
+	stmt = `UPDATE ocr2dr_requests SET state=$3 WHERE request_id=$1 AND contract_address=$2;`
```

Comment:
What I was asking here in my earlier comment is whether this UPDATE and earlier SELECT (line 79) should happen in a single transaction so that data doesn't change in between those calls?

Reply:
That's a great point. I will wrap those into a transaction. Thank you!

Reply:
Done.